### DISCUSSION TITLE
Re-project match with a large raster

### DISCUSSION THREAD

**Discussion by JimShady** [corteva/rioxarray#757] [Q&A] [answered]
I am trying to reproject a large raster, using a second raster as a template. This works fine on smaller rasters, but crashes with larger ones. I think I probably need to use window'd read and writes but I'm unsure how to implement this or if it's the correct solution. Thanks.
    
    my_raster = rioxarray.open_rasterio("large.tif")

    reference = rioxarray.open_rasterio("template_large_tif")

    reprojected = my_raster.rio.reproject_match(reference)

    reprojected.rio.to_raster("output.tif")

**Comment by snowman2** [comment, marked as answer]
Re-project match just uses the same CRS, transform, and shape of the other raster:
https://github.com/corteva/rioxarray/blob/6b768ec260fc65752c5396fbfe3a9b25e49ae2b1/rioxarray/raster_array.py#L552

Examples referenced in solution:
- [re-project a large raster](https://corteva.github.io/rioxarray/stable/examples/reproject.html#Reproject-Large-Rasters-with-Virtual-Warping)
- [reading files](https://corteva.github.io/rioxarray/stable/examples/read-locks.html#Caching-Considerations)
- [writing files](https://corteva.github.io/rioxarray/latest/examples/convert_to_raster.html#Memory-efficient-raster-writing)

Solution:

```python

import rasterio
import rasterio.vrt
import rioxarray

with rasterio.open("large.tif") as src, rasterio.open("template_large.tif") as template:
        with rasterio.vrt.WarpedVRT(
            src,
            crs=template.crs,
            transform=template.transform,
            height=template.height,
            width=template.width,
        ) as vrt:
            rds = rioxarray.open_rasterio(vrt, lock=False, parse_coordinates=False)
            rds.rio.to_raster("output.tif", tiled=True, windowed=True)
```

**Reply by JimShady** [reply]
Thanks for the help @snowman2  . Not very experienced with this library so struggling to implement your suggestion a bit. Looking at the URL you gave

```
env = rasterio.Env(
    GDAL_DISABLE_READDIR_ON_OPEN="EMPTY_DIR",
    CPL_VSIL_CURL_USE_HEAD=False,
    CPL_VSIL_CURL_ALLOWED_EXTENSIONS="TIF",
)
with env:
    with rasterio.open(url) as src: #this would be the file I want to reproject
        with rasterio.vrt.WarpedVRT(src, crs="EPSG:4326") as vrt: # I guess this is warping my file to 4326 (can I specify a template file here instead?)
            rds = rioxarray.open_rasterio(vrt) # This is opening the result I think. But I want the result written to file?
```

**Reply by JimShady** [reply]
i.e. I can't see from the example how I can use a 'template' raster, and/or how to write out the result.

**Reply by JimShady** [reply]
I used an AI bot to come up with this. Going to test it now.

```
output = "output.tif"

raster = rasterio.open("input_file.tif")
reference = rasterio.open("template.tif")

# Get the chunk size based on the raster dimensions
chunk_size = (raster.height // 10, raster.width // 10)

# Reproject the raster in smaller chunks
with rasterio.open(output , 'w', **raster.profile) as dst:
      dst.write(
          rasterio.warp.reproject(
              source=raster.read(),
              destination=dst,
              src_transform=raster.transform,
              src_crs=raster.crs,
              dst_transform=reference.transform,
              dst_crs=reference.crs,
              resampling=Resampling.bilinear,
              num_threads=1,
              chunk_size=chunk_size
          )
      )
```

**Reply by snowman2** [reply]
@JimShady I added a code example to reference.

**Reply by JimShady** [reply]
Thanks so much @snowman2  . Will give it a whirl over the weekend.

**Comment by RichardScottOZ** [comment]
reproject match is like this

```python

template_raster = rioxarray.open_rasterio('template_raster.tif')

useraster = rioxarray.open_rasterio('useraster.tif')

finalraster = useraster.rio.reproject_match(template_raster)

finalraster = finalraster.assin_coords({"x":template_raster.x, "y":template_raster.y})   #so they match exactly
```

**Reply by JimShady** [reply]
Thanks @RichardScottOZ  . Unfortunately I've still not managed to figure out how to do this with large rasters. I think I'm going to resort to 'raw' gdal on the command line via os.system() commands.

**Comment by Curallin** [comment]
This is GPT's answer, I would like to know why the code execution to reprojected_ds only takes 17.3s, however using rio.to_raster() over 40 minutes still for completion, is it because my parameter setting is not right?
```
import dask.array as da
import dask
from odc.geo.xr import xr_reproject
import rioxarray  

input_raster = r'D:\Geographical_data\land_use\CLCD_v01_2023_albert.tif'
reproj_raster = r'D:\Geographical_data\land_use\CLCD_v01_2023_WGS1984_4_128.tif'

dst_crs = 'EPSG:4326'

dask.config.set(scheduler='threads', num_workers=64)

ds = rioxarray.open_rasterio(input_raster, chunks={'x': 3000, 'y': 3000})

reprojected_ds = xr_reproject(
    src=ds,
    how=dst_crs, 
    resampling='nearest',  
    resolution='fit',  
    dst_nodata='auto', 
    shape=None,  
    tight=False  
)

delayed_write = reprojected_ds.rio.to_raster(
    reproj_raster,
    compress='LZW',  
    dtype=ds.dtype,  
    windowed=True,  
    lock=None, 
    compute=False 
)

dask.compute(delayed_write)
```
